### PR TITLE
Tidy scipp imports

### PR DESCRIPTION
**Pull request by nvaytet**
Instead of importing all the cpp functions (binded through `pybind11`), we only import the wrapper functions.

In addition:
- renamed `units_neutron.cpp` to `units.cpp`
- moved the top level `rotation_matrix_from_quaternion_coeffs` function into an `eigen` submodule (this function is a bit of an oddity anyway, maybe it should be moved into the mantid compatibility module of `scippneutron`?)

### CODE REVIEW

File: python/eigen.cpp [hunk 1]
```diff
@@ -11,7 +11,8 @@
 namespace py = pybind11;
 
 void init_eigen(py::module &m) {
-  m.def(
+  auto eigen_m = m.def_submodule("eigen");
+  eigen_m.def(
```
[the remark this answers is not in the document]
Comment (SimonHeybrock) [reply python/eigen.cpp:15]: Make it part of `geometry` maybe?

**Review by SimonHeybrock — APPROVED** [review]

**Comment by SimonHeybrock** [comment]
Did you check if `dir(sc)` looks reasonable now?

**Comment by nvaytet** [comment]
> Did you check if `dir(sc)` looks reasonable now?

Here is the list of names that have disappeared:
```
BinEdgeError
BinnedDataError
CoordsConstView
CoordsView
CoordsView_items_view
CoordsView_keys_view
CoordsView_values_view
DataArrayConstView
DataArrayError
DatasetConstView
DatasetError
DatasetView_items_view
DatasetView_keys_view
DatasetView_values_view
Dataset_items_view
Dataset_keys_view
Dataset_values_view
DimensionError
ElementArrayView_DataArray
ElementArrayView_Dataset
ElementArrayView_Eigen_Matrix3d
ElementArrayView_Eigen_Vector3d
ElementArrayView_Variable
ElementArrayView_bool
ElementArrayView_bucket_DataArray
ElementArrayView_bucket_Dataset
ElementArrayView_bucket_Variable
ElementArrayView_double
ElementArrayView_float
ElementArrayView_int32
ElementArrayView_int64
ElementArrayView_string
MasksConstView
MasksView
MasksView_items_view
MasksView_keys_view
MasksView_values_view
NotFoundError
SizeError
Slice
SliceError
VariableConstView
VariableError
VariancesError
bin_size
bin_with_coords
bins_begin_end
bins_data
bins_dim
divide
floor_divide
get_slice_params
is_bins
logical_and
logical_or
logical_xor
minus
mod
plus
rotation_matrix_from_quaternion_coeffs
span_DataArray
span_Dataset
span_Dim_const
span_Eigen_Matrix3d
span_Eigen_Vector3d
span_Variable
span_bool
span_double
span_double_const
span_float
span_long
span_long_const
span_string
span_string_const
times
```

**Comment by nvaytet** [comment]
Looks a bit inconsistent with the errors (`NotFoundError`, `SizeError`, etc)
I had to import a few to get the tests to pass, but maybe I should import all of them?
Or import none of them and find another way to fix the tests?
It's tests that have something like `with pytest.raises(sc.CoordError):`

**Comment by SimonHeybrock** [comment]
> Looks a bit inconsistent with the errors (`NotFoundError`, `SizeError`, etc)
> I had to import a few to get the tests to pass, but maybe I should import all of them?
> Or import none of them and find another way to fix the tests?
> It's tests that have something like `with pytest.raises(sc.CoordError):`

Yes, we absolutely need to import those errors. The whole point of different exception types is so the user can catch exceptions selectively. Otherwise we could have just throw `RuntimeError` everywhere.

**Comment by SimonHeybrock** [comment]
> logical_xor

This and other free functions numpy-like also need to be imported of course.

**Comment by SimonHeybrock** [comment]
@nvaytet It would be worth updating this now that `dev` has been merged into `main`.

**Comment by nvaytet** [comment]
@SimonHeybrock do these also need importing?
```
bin_size
bin_with_coords
bins_begin_end
bins_data
bins_dim
```
and what about
```
minus
mod
plus
etc..
```
?

**Comment by SimonHeybrock** [comment]
If numpy has the function we should import it. We need to fix naming of some of our ops...

**Comment by nvaytet** [comment]
yes, `minus` is called `subtract` in numpy

File: python/src/scipp/__init__.py [hunk 2]
```diff
@@ -22,8 +22,21 @@ def custom_formatwarning(msg, *args, **kwargs):
         'You are running a "Debug" build of scipp. For optimal performance use a "Release" build.'
     )
 
-from ._scipp.core import *
 from ._scipp import __version__
+# Import classes
+from ._scipp.core import Variable, DataArray, Dataset, GroupByDataArray, \
+                         GroupByDataset, Dim
```
Comment (SimonHeybrock) [inline python/src/scipp/__init__.py:28]: Do we need `Dim`?

Comment (nvaytet) [reply python/src/scipp/__init__.py:28]: There is à `test_dim.py` file but I could just remove those tests

Comment (SimonHeybrock) [reply python/src/scipp/__init__.py:28]: I don't see why not.

**Comment by nvaytet** [comment]
I don't understand why i am seeing `Any` and `Union` from `typing` in `dir(sc)`
Screenshot at 2021-04-28 13-44-55

**Comment by SimonHeybrock** [comment]
> I don't understand why i am seeing `Any` and `Union` from `typing` in `dir(sc)`
> Screenshot at 2021-04-28 13-44-55

Because of `from _bla.py import *` in our `__init__.py`?

**Comment by nvaytet** [comment]
Yes, in `_variable.py` for instance.
Btu where is the good balance between having a clean `sc` module when imported and convenience: not having to
`from ._variable import array, linspace, zeros, zeros_like, logspace....`
?

**Comment by SimonHeybrock** [comment]
People may want to run `dir(sc)` to see what is there.

**Comment by nvaytet** [comment]
`dir(sc)` now gives
```
['BinEdgeError',
 'BinnedDataError',
 'Bins',
 'CoordError',
 'DTypeError',
 'DataArray',
 'DataArrayError',
 'Dataset',
 'DatasetError',
 'DimensionError',
 'GroupByDataArray',
 'GroupByDataset',
 'GroupbyBins',
 'NotFoundError',
 'SizeError',
 'SliceError',
 'Unit',
 'UnitError',
 'Variable',
 'VariableError',
 'VariancesError',
 '__builtins__',
 '__cached__',
 '__doc__',
 '__file__',
 '__loader__',
 '__name__',
 '__package__',
 '__path__',
 '__spec__',
 '__version__',
 '_bins',
 '_comparison',
 '_counts',
 '_cpp_wrapper_util',
 '_cumulative',
 '_dataset',
 '_debug_',
 '_groupby',
 '_groupby_bins',
 '_make_sizes',
 '_math',
 '_obj',
 '_operations',
 '_reduction',
 '_repr_html_',
 '_scipp',
 '_set_bins',
 '_shape',
 '_to_hdf5',
 '_trigonometry',
 '_unary',
 '_utils',
 '_variable',
 'abs',
 'acos',
 'all',
 'any',
 'arange',
 'array',
 'asin',
 'atan',
 'atan2',
 'bin',
 'bins',
 'broadcast',
 'buckets',
 'choose',
 'collapse',
 'combine_masks',
 'compat',
 'concatenate',
 'config',
 'cos',
 'counts_to_density',
 'cumsum',
 'datetime64',
 'density_to_counts',
 'divide',
 'dot',
 'dtype',
 'empty',
 'empty_like',
 'equal',
 'exp',
 'extend_units',
 'filter',
 'flatten',
 'floor_divide',
 'fold',
 'from_dict',
 'geometry',
 'geomspace',
 'greater',
 'greater_equal',
 'groupby',
 'histogram',
 'html',
 'identical',
 'io',
 'is_data_array',
 'is_dataset',
 'is_dataset_or_array',
 'is_variable',
 'isclose',
 'isfinite',
 'isinf',
 'islinspace',
 'isnan',
 'isneginf',
 'isposinf',
 'issorted',
 'less',
 'less_equal',
 'linspace',
 'log',
 'log10',
 'logical_and',
 'logical_or',
 'logical_xor',
 'logspace',
 'lookup',
 'make_html',
 'make_svg',
 'max',
 'mean',
 'merge',
 'min',
 'minus',
 'mod',
 'nan_to_num',
 'nanmax',
 'nanmean',
 'nanmin',
 'nansum',
 'norm',
 'not_equal',
 'object_list',
 'ones',
 'ones_like',
 'plot',
 'plotting',
 'plus',
 'rebin',
 'reciprocal',
 'reshape',
 'scalar',
 'show',
 'sin',
 'sizes',
 'slices',
 'sort',
 'split',
 'sqrt',
 'stddevs',
 'sum',
 'table',
 'tan',
 'times',
 'to_dict',
 'to_html',
 'to_unit',
 'transpose',
 'units',
 'user_configuration_filename',
 'values',
 'variances',
 'zeros',
 'zeros_like']
```

**Review by SimonHeybrock — APPROVED** [review]

**Pull request merged**